File: 2015/12/rust/src/main.rs
use regex::Regex;
use serde_json::Value;

// --- Constants ---
const INPUT_FILE:&'static str = "../input/input.txt";
const PARSER_REGEX: &'static str = r"(-?\d+)";

// --- Sum all number of a json, ignoring object with "red" attribute ---
fn sum(json: &Value) -> isize {
    match json {
        Value::Number(n) => n.as_i64().unwrap().try_into().unwrap(),
        Value::Array(a)  => a.iter().map(sum).sum(),
        Value::Object(o) => {
            for v in o.values() {
                if v.as_str() == Some("red") { return 0; }
            }
            o.values().map(sum).sum()
        },
        _ => 0
    }
}

// --- Main function ---
fn main() -> Result<(), Box<dyn std::error::Error>> {
    // --- Read and parse the input file ---
    let text = std::fs::read_to_string(INPUT_FILE)?;

    // --- Puzzle 1 ---
    let regex = Regex::new(PARSER_REGEX)?;
    let total: isize = regex.captures_iter(&text)
        .map(|x| x[1].parse::<isize>().unwrap())
        .sum();
    println!("1. Sum of all numbers: {}", total);

    // --- Puzzle 2 ---
    let json: Value = serde_json::from_str(&text)?;
    let total = sum(&json);
    println!("2. Sum of all numbers: {}", total);

    Ok(())
}

File: 2015/12/go/main.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
	"strings"
)

// --- Constants ---
const FILENAME = "../input/input.txt"
const REGEX = `(-?\d+)`

func sum(jsonValue interface{}) int {
	switch v := jsonValue.(type) {
	case float64:
		return int(v)
	case []interface{}:
		total := 0
		for _, x := range v {
			total += sum(x)
		}
		return total
	case map[string]interface{}:
		for _, x := range v {
			if str, ok := x.(string); ok && str == "red" {
				return 0
			}
		}
		total := 0
		for _, x := range v {
			total += sum(x)
		}
		return total
	default:
		return 0
	}
}

func main() {
	// --- Variables ---
	re := regexp.MustCompile(REGEX)

	// --- Read input file ---
	bytes, err := os.ReadFile(FILENAME)
	if err != nil {
		log.Fatal(err)
	}
	text := strings.TrimSpace(string(bytes))

	// --- Puzzle 1 ---
	total := 0
	for _, c := range re.FindAllString(text, -1) {
		n, err := strconv.Atoi(c)
		if err != nil {
			log.Panic(err)
		}
		total += n
	}
	fmt.Println("1. Sum of all numbers:", total)

	// --- Puzzle 2 ---
	var data interface{}
	err = json.Unmarshal([]byte(text), &data)
	if err != nil {
		log.Panic(err)
	}
	total = sum(data)
	fmt.Println("2.Sum of all numbers:", total)
}
